doc/smart-switch/smart-switch-database-architecture/smart-switch-database-design.md

```diff
@@ -0,0 +1,415 @@
+    "delayed": "False",
+    "has_global_scope": "True",
+    "has_per_asic_scope": "True",
+    "has_per_dpu_scope": "True", # New field for DPU database service
+    "high_mem_alert": "disabled",
+    "state": "always_enabled",
+    "support_syslog_rate_limit": "true"
+},
+```
+
+Our design also extends its multi-ASIC principles by introducing a database_global.json file. It includes two critical fields:
+
+- database_name: This field uniquely maps to the DPU's index, ensuring a clear association between the database instances and the respective DPU.
+- include: The include field serves as a pointer to the location of the DPU's database configuration.
+
+```json
+{
+  "INCLUDES": [
+    {
+      "include": "../../redis/sonic-db/database_config.json"
+    },
+    {
+      "database_name": "dpu1",
+      "include": "../../redisdpu0/sonic-db/database_config.json"
+    },
+    {
+      "database_name": "dpu0",
+      "include": "../../redisdpu1/sonic-db/database_config.json"
+    }
+  ],
+  "VERSION": "1.0"
+}
+```
+
+Within the NPU, the management of DPU overlay databases involves specific configurations. Each DPU overlay database instance is bound to the IP address of the midplane bridge (169.254.200.254 by default). The TCP port assignment follows a predictable pattern, with each DPU ID associated with a unique port (6381 + DPU ID).
+
+Here is an example includes two DPU:
+
+```json
+# DPU0: /var/run/redisdpu2/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6381,
+    "unix_socket_path": "/var/run/redisdpu0/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+#DPU1: /var/run/redisdpu1/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6382,
+    "unix_socket_path": "/var/run/redisdpu1/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+```
+
+There are four new tables introduction for the DPU overlay database:
+
+```json
+"DPU_APPL_DB": {
+    "id": 15,
+    "separator": ":",
+    "instance": "redis",
+    "format": "proto"
+},
+"DPU_APPL_STATE_DB": {
+    "id": 16,
+    "separator": "|",
+    "instance": "redis"
+},
+"DPU_STATE_DB": {
+    "id": 17,
+    "separator": "|",
+    "instance": "redis"
+},
+"DPU_COUNTERS_DB": {
+    "id": 18,
+    "separator": ":",
+    "instance": "redis"
+}
+```
+
+##### DPU side
+
+In the architecture of our Smart Switch, DPU operation involves accessing both local and remote database services. This document is on elucidating the interaction with remote database services in the NPU, specifically for overlay objects.
+
+The DPU employs a DHCP server hosted on the NPU, ensuring each DPU fetches a predetermined and consistent IP address. Leveraging this IP address, the DPU determines the TCP port of its associated overlay database. Following this design principle, the DPU autonomously generates the requisite configuration for remote database services.
```

Comment:
We might have a problem here with the boot order of the systemd services. The database service is one of the first services that start after the system boot. To start the database service we need to have `database_config.json` file available. However, to generate the file we need to have configuration from the DHCP server. The configuration can be received from the DHCP server only after the start of the `interface-config.service` which starts later than the database service. So with the existing boot order, this can't be done. 

I think to proceed with this we need to have a flow diagram for the `database_config.json` file generation to better understand all the dependencies.

Reply:
Thanks for your suggestion, I will check that.

Reply:
For the further discussion in PR: https://github.com/sonic-net/SONiC/pull/1586